File: GOLD/closingthefarm.py
class dsu:
    def __init__(self, size: int):
        self.parent = [i for i in range(size)]
        self.sizes = [1 for _ in range(size)]
        self.size = size
    
    def find_parent(self, a: int) -> int:
        if (a == self.parent[a]):
            return a
        self.parent[a] = self.find_parent(self.parent[a])
        return self.parent[a]

    def union_set(self, a: int, b: int):
        a = self.find_parent(a)
        b = self.find_parent(b)
        if (a != b):
            # a is not in the same set as b
            if (self.sizes[a] < self.sizes[b]):
                a,b = b,a
            
            # perform union
            self.parent[b] = a
            self.sizes[a] = self.sizes[b]

            self.size -= 1

fin = open("closing.in", "r")
fout = open("closing.out", "w")

n, m = [int(x) for x in fin.readline().strip().split(" ")]
d = dsu(n)

adj = []
for _ in range(n):
    adj.append([])

for _ in range(m):
    f,t = [int(x)-1 for x in fin.readline().strip().split(" ")]
    adj[f].append(t)
    adj[t].append(f)

order = []
for i in range(n):
    order.append(int(fin.readline().strip())-1)
order = reversed(order)


exists = [False for _ in range(n)]
ans = [False for _ in range(n)]

num = 0 # number of nodes placed
for i in order:

    num += 1
    exists[i] = True

    for x in adj[i]:
        # check if the other node exists
        # if it does exist => add the edge
        if (exists[x]):
            d.union_set(i, x)
    
    # n - num => position in ans array
    # d.size - (n-num) => need to subtract out nodes that don't exist.
    ans[n-num] = (d.size - (n-num) == 1)

for i in ans:
    if (i):
        fout.write("YES\n")
    else:
        fout.write("NO\n")

File: GOLD/closingthefarm.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

//dsu data structure 
class dsu {
    public:
    vector<int> parent;
    vector<int> s;
    int size = 0;
    dsu(int n) {
        parent.resize(n);
        s.resize(n, 1);
        size = n;
        for (int i = 0 ; i < n ; i++) parent[i] = i;
    }

    int find_set(int a) {
        if (a == parent[a])
            return a;
        return parent[a] = find_set(parent[a]);
    }

    void union_set(int a, int b) {
        a = find_set(a);
        b = find_set(b);
        if (a != b) {
            if (s[a] < s[b]) {
                int temp = a;
                a = b;
                b = temp;
            }
            parent[b] = a;
            s[a] += s[b];
            //cout << a << " " << b << endl;
            size--;
        }
    }
};

int main() {
    int n,m;
    cin >> n >> m;
    dsu d(n);

    vector<vector<int>> v(n);
    vector<pair<int,int>> p(m);
    for (int i = 0 ; i < m ; i++) {
        /*
        int f,t;
        cin >> f >> t;
        f--;t--;
        v[f].push_back(t);
        v[t].push_back(f);
        */
        cin >> p[i].first >> p[i].second;
        p[i].first--; p[i].second--;
    }

    vector<int> o(n);
    vector<int> t(n);
    for (int i = 0 ; i < n ; i++) {
        cin >> o[i];
        o[i]--;
        t[o[i]] = i;
    }

    for (int i = 0 ; i < m ; i++) {
        if (t[p[i].first] > t[p[i].second]) {
            v[p[i].second].push_back(p[i].first);
        } else {
            v[p[i].first].push_back(p[i].second);
        }
    }

    vector<bool> ans(n);
    int k = 0;
    for (int i = n-1; i >= 0 ; i--) {
        k++;
        for (int j = 0 ; j < v[o[i]].size() ; j++) {
            if (d.find_set(o[i]) != d.find_set(v[o[i]][j])) {
                d.union_set(o[i], v[o[i]][j]);
                k--;    
            }
        }
        ans[i] = k <= 1;
    }
    for (int i = 0 ; i < n ; i++) {
        if (ans[i]) {
            cout << "YES" << endl;
        } else {
            cout << "NO" << endl;
        }
    }
}